src/contracts/atlas/Sorter.sol

```diff
@@ -0,0 +1,199 @@
+        if (searcherLastActiveBlock >= block.number) {
+            return false;
+        }
+
+        // Make sure the searcher nonce is accurate
+        uint256 nextSearcherNonce = IEscrow(escrow).nextSearcherNonce(searcherCall.metaTx.from);
+        if (nextSearcherNonce != searcherCall.metaTx.nonce) {
+            return false;
+        }
+
+        // Make sure that the searcher has the correct codehash for protocol control contract
+        if (protocolCall.to.codehash != searcherCall.metaTx.controlCodeHash) {
+            return false;
+        }
+
+        // Make sure that the searcher's maxFeePerGas matches or exceeds the user's
+        if (searcherCall.metaTx.maxFeePerGas < userMetaTx.maxFeePerGas) {
+            return false;
+        }
+
+        return true;
+    }
+
+    function _getSortingData(
+        ProtocolCall memory protocolCall, 
+        UserCall calldata userCall, 
+        SearcherCall[] calldata searcherCalls,
+        uint256 count
+    ) internal view returns (SortingData[] memory, uint256){
+
+        BidData[] memory bidFormat = IProtocolControl(protocolCall.to).getBidFormat(userCall.metaTx);
+
+        SortingData[] memory sortingData = new SortingData[](count);
+
+        uint256 i;
+        uint256 invalid;
+        for (;i<count;) {
+            if (
+                _verifyBidFormat(bidFormat, searcherCalls[i]) && 
+                _verifySearcherEligibility(protocolCall, userCall.metaTx, searcherCalls[i])
+            ) {
+                sortingData[i] = SortingData({
+                    amount: IProtocolControl(protocolCall.to).getBidValue(searcherCalls[i]),
+                    valid: true
+                });
+                
+
+            } else {
+                sortingData[i] = SortingData({
+                    amount: 0,
+                    valid: false
+                });
+                unchecked{ ++invalid; }
+            }
+            unchecked{ ++i; }            
+        }
+
+        return (sortingData, invalid);
+    }
+
+    function _sort(
+        SortingData[] memory sortingData,
+        uint256 count,
+        uint256 invalid
+    ) internal pure returns (uint256[] memory) {
+
+        uint256[] memory sorted = new uint256[](count - invalid);
+
+        uint256 n; // outer loop counter
+        uint256 i; // inner loop counter
+
+        uint256 topBid;
+        uint256 bottomBid;
+
+        for (;invalid<count;) {
+
+            // Reset the ceiling / floor
+            topBid = 0;
+            bottomBid = type(uint256).max;
+
+            // Loop through and find the highest and lowest remaining valid bids
+            for(;i<sortingData.length;) {
+                if (sortingData[i].valid) {
+                    if (sortingData[i].amount > topBid) {
+                        sorted[n] = i;
+                    }
+                    if (sortingData[i].amount < bottomBid) {
+                        sorted[count-1-n] = i;
```

Comment:
Shouldn't this line also take invalid into account?
```suggestion
                        sorted[count-invalid-1-n] = i;
```

Reply:
great catch.